include/pybind11/pytypes.h

```diff
@@ -298,6 +298,22 @@ class handle : public detail::object_api<handle> {
 
 #ifdef PYBIND11_HANDLE_REF_DEBUG
 private:
+    void throw_gilstate_error(const std::string &function_name) const {
+        fprintf(stderr,
+                "%s is being called while PyGILState_Check() is failing. "
+                "This usually indicates that you are calling pybind11 functions without holding "
+                "the GIL or before "
+                "Python (and/or this Python module) is finished initializing.\n",
+                function_name.c_str());
+        if (Py_TYPE(m_ptr)->tp_name != nullptr) {
```

Comment:
Is Py_TYPE(m_ptr safe without a GIL?

Reply:
No. That's why I have two different fprintf statements there. So if Py_TYPE dies, we at least get the first fprintf.

Reply:
Good question @Skylion007 

I also think there isn't anything meaningful we can do here other than trying and crashing in the worst case.

`Py_TYPE` is a really simple macro, it will most likely work, unless some other thread deletes the type object concurrently.

Reply:
The only other thing we could do here that would be good would be to grab an actual stacktrace.

The issue is that grabbing stacktraces in a platform portable way is a nightmare.

Reply:
Yes. It's a thankless tedious project. Let's not go there.